File: backend/routes/hierarchy.py
"""
backend/routes/hierarchy.py -- Hierarchy Search endpoints. Two endpoints
because step 3's drill-down is per-video and re-triggerable independently
(seed-frame change, "Expand" button) without re-running step 1 -- the
frontend caches each group's step-1 `frames` list from the initial search
response and resends it verbatim on a later expand call
(`hierarchy_expand_group` always drills down from the ORIGINAL step-1
list, never a previously-drilled/compounded one).
"""

from typing import List, Optional

from fastapi import APIRouter
from pydantic import BaseModel

from .. import config
from ..core.query import resolve_query
from ..filters.lot import parse_lot_range
from ..search.composite import hierarchy as hier_mod
from .schemas import SearchScope

router = APIRouter()


class HierarchySearchRequest(SearchScope):
    # No od_filter: Hierarchy's drill-down is a picture query per video.
    query: Optional[str] = None
    image_id: Optional[str] = None
    top_g: int = config.TOP_G_DEFAULT


class HierarchyGroup(BaseModel):
    video_id: str
    best_rank: int
    score_label: str
    best_score_val: float
    step1_frames: list
    seed_options: List[int]
    top1_n: int
    drilled_frames: list


class HierarchySearchResponse(BaseModel):
    groups: List[HierarchyGroup] = []


@router.post("/api/search/hierarchy", response_model=HierarchySearchResponse)
def search_hierarchy(body: HierarchySearchRequest):
    query = resolve_query(body.query, body.image_id)
    fetch_k = max(config.FETCH_K, body.top_k)
    lot_filter = parse_lot_range(body.lot_filter, body.exclude_lot)

    raw_groups = hier_mod.base_search_grouped(
        query, fetch_k, body.video_filter, lot_filter, body.top_k, body.facet_field, body.facet_value)

    groups = []
    for g in raw_groups:
        frames = g["frames"]
        best = frames[0]
        seed_options = [f["n"] for f in frames]
        top1_n = seed_options[0]
        drilled = hier_mod.hierarchy_expand_group(g["video_id"], frames, body.top_g, fetch_k, seed_n=top1_n)
        groups.append(HierarchyGroup(
            video_id=g["video_id"], best_rank=best["rank"], score_label=best["score_label"],
            best_score_val=best["score_val"], step1_frames=frames, seed_options=seed_options,
            top1_n=top1_n, drilled_frames=drilled,
        ))
    return HierarchySearchResponse(groups=groups)


class HierarchyExpandRequest(BaseModel):
    video_id: str
    step1_frames: list
    top_g: int = config.TOP_G_DEFAULT
    seed_n: Optional[int] = None
    top_k: int = config.DISPLAY_N


class HierarchyExpandResponse(BaseModel):
    frames: list = []


@router.post("/api/search/hierarchy/expand", response_model=HierarchyExpandResponse)
def expand_hierarchy(body: HierarchyExpandRequest):
    fetch_k = max(config.FETCH_K, body.top_k)
    frames = hier_mod.hierarchy_expand_group(body.video_id, body.step1_frames, body.top_g, fetch_k, seed_n=body.seed_n)
    return HierarchyExpandResponse(frames=frames)

File: frontend/js/signals/hierarchy.js
// frontend/js/signals/hierarchy.js -- Hierarchy Search: bespoke rendering
// (not renderGrid) -- a video-grouped, drilled-down result set isn't a
// plain ranked list.

import { searchHierarchy, expandHierarchy } from "../core/api.js";
import { renderThumb, renderActions } from "../ui/render.js";
import { currentQuery } from "../ui/query-input.js";
import { hierExtraG, resetExportCandidates, scopeFilters } from "../core/state.js";
import { setGroupByUi, tile } from "../core/settings.js";

const topGWrap = document.getElementById("top-g-wrap");

// video_id -> {best, step1_frames, seed_n} -- cached so the seed-picker and
// "Expand" button can re-drill without re-running Step 1's base search,
// always from the original Step-1 list rather than a previously-drilled
// one (see backend/routes/search/hierarchy.py's module docstring).
let groupsCache = new Map();

export function mount(controlsEl) {
    controlsEl.innerHTML = `<div class="thumb-caption muted">SigLIP2 frame search, grouped by video, drilled down to Top-G frames/video. No leg choice -- text or picture query, SigLIP2 only.</div>`;
    topGWrap.style.display = "block";
    setGroupByUi({ visible: false }); // Hierarchy is always grouped by video by construction
}

export function unmount() {
    topGWrap.style.display = "none";
    setGroupByUi();
}

async function renderGroupFrames(vid, container) {
    const g = groupsCache.get(vid);
    const grid = container.querySelector(".hier-frame-grid");
    grid.innerHTML = "";
    // Per-frame actions, same as render.js's grouped renderGrid() -- the
    // group only controls layout, each frame still needs its own targets.
    for (const r of g.drilled_frames) {
        const cell = renderThumb(r);
        cell.append(renderActions(r.video_id, r.n));
        grid.append(cell);
    }
}

async function redrill(vid, container) {
    const g = groupsCache.get(vid);
    const topK = parseInt(document.getElementById("top-k").value, 10) || 200;
    const topG = (parseInt(document.getElementById("top-g").value, 10) || tile().topG) + (hierExtraG.get(vid) || 0);
    const data = await expandHierarchy({
        video_id: vid, step1_frames: g.step1_frames, top_g: topG, seed_n: g.seed_n, top_k: topK,
    });
    g.drilled_frames = data.frames;
    await renderGroupFrames(vid, container);
}

function renderGroup(container, vid) {
    const g = groupsCache.get(vid);
    container.innerHTML = "";

    const header = document.createElement("div");
    header.className = "group-header";
    header.innerHTML = `<b>${vid}</b> · best rank ${g.best_rank} · ${g.score_label}=${g.best_score_val.toFixed(4)} · ${g.step1_frames.length} frame(s) from Step 1`;
    container.append(header);

    // Expand stays video-level (it pulls more frames for this one video --
    // there's no per-frame equivalent), separate from the per-frame
    // show-more/play/copy/export buttons now on each drilled frame below.
    // Step follows the tile-size setting (settings.js's TILE_SIZES), same as
    // the Top-G default it adds onto: one more row of frames per click.
    const expandStep = tile().hierExpand;
    const expandBtn = document.createElement("button");
    expandBtn.className = "btn";
    expandBtn.title = `Pull in ${expandStep} more frames from this video`;
    expandBtn.textContent = "⬇ Expand";
    expandBtn.style.margin = "0.3rem 0";
    expandBtn.onclick = () => {
        hierExtraG.set(vid, (hierExtraG.get(vid) || 0) + expandStep);
        redrill(vid, container);
    };
    container.append(expandBtn);

    const seedSelect = document.createElement("select");
    seedSelect.style.cssText = "width:100%;margin:0.3rem 0;padding:0.35rem;border:1px solid var(--border);border-radius:6px;";
    for (const n of g.seed_options) {
        const opt = document.createElement("option");
        opt.value = n;
        opt.textContent = `Seed: frame ${n}` + (n === g.top1_n ? " (top-1)" : "");
        if (n === g.seed_n) opt.selected = true;
        seedSelect.append(opt);
    }
    seedSelect.onchange = () => {
        g.seed_n = parseInt(seedSelect.value, 10);
        redrill(vid, container);
    };
    container.append(seedSelect);

    const grid = document.createElement("div");
    grid.className = "grid hier-frame-grid";
    container.append(grid);

    const hr = document.createElement("hr");
    hr.className = "divider";
    container.append(hr);

    renderGroupFrames(vid, container);
}

export async function run(resultsEl, statusEl) {
    const { query, image_id } = currentQuery();
    if (!query.trim() && !image_id) {
        resultsEl.innerHTML = "";
        statusEl.innerHTML = `<div class="status-banner info">Type a query, or paste an image, to search.</div>`;
        return;
    }
    statusEl.innerHTML = "";

    const topK = parseInt(document.getElementById("top-k").value, 10) || 200;
    const topG = parseInt(document.getElementById("top-g").value, 10) || tile().topG;
    const body = { query: image_id ? null : query, image_id, top_k: topK, top_g: topG, ...scopeFilters() };

    resultsEl.innerHTML = `<div class="status-banner info">Searching…</div>`;
    let data;
    try {
        data = await searchHierarchy(body);
    } catch (e) {
        resultsEl.innerHTML = `<div class="status-banner error">${e.message}</div>`;
        return;
    }
    resultsEl.innerHTML = "";

    if (!data.groups.length) {
        resultsEl.innerHTML = `<div class="status-banner info">No results.</div>`;
        return;
    }
    // Flatten to one candidate per video (its Step 1 top-1 frame, the same
    // frame renderActions()'s export (★) button targets below) -- Hierarchy's
    // grouped/drilled-down shape isn't a plain ranked list, but the export
    // popup's "similars" preview needs one.
    resetExportCandidates(data.groups.map((g) => ({
        video_id: g.video_id, n: g.step1_frames[0].n, rank: g.best_rank,
        score_label: g.score_label, score_val: g.best_score_val, text: null,
    })));

    groupsCache = new Map();
    const h = document.createElement("h2");
    h.textContent = "Hierarchy Search";
    resultsEl.append(h);

    for (const g of data.groups) {
        groupsCache.set(g.video_id, {
            best_rank: g.best_rank, score_label: g.score_label, best_score_val: g.best_score_val,
            best_n: g.step1_frames[0].n, step1_frames: g.step1_frames, seed_options: g.seed_options,
            top1_n: g.top1_n, seed_n: g.top1_n, drilled_frames: g.drilled_frames,
        });
        const groupEl = document.createElement("div");
        resultsEl.append(groupEl);
        renderGroup(groupEl, g.video_id);
    }
}
